sqlitecluster/SQLiteNode.cpp

```diff
@@ -1371,22 +1559,53 @@ bool SQLiteNode::update(uint64_t& nextActivity) {
                             SASSERTWARN(_currentCommand->transaction.empty());
                             SINFO("Starting processing command '" << _currentCommand->request.methodLine << "' ("
                                                                   << _currentCommand->id << ")");
-                            _processCommandWrapper(_db, _currentCommand);
+
+                            bool unlock = false;
+                            if (_currentCommand->processed >= MAX_PROCESS_TRIES) {
+                                SWARN("[concurrent] Command " << _currentCommand->id << " processed " << _currentCommand->processed
+                                      << " times, exceeded max of " << MAX_PROCESS_TRIES
+                                      << ", forcing synchronous commit.");
+                                unlock = true;
+                                SQLite::commitLock.lock();
+                            }
+                            bool needsCommit = _processCommandWrapper(_db, _currentCommand);
+
                             SASSERT(!_currentCommand->response.empty()); // Must set a response
 
                             // Anything to commit?
-                            if (_db.insideTransaction()) {
+                            if (needsCommit) {
+
+                                // We're about to send a transaction here, grab our commit mutex and send anything
+                                // outstanding. This remains locked until we've finished this transaction.
+                                SQLite::commitLock.lock();
```

Comment:
True, but it doesn't look to me like it unlocks correctly -- I'm having trouble following the logic, but it seems like it will lock it more times than it unlocks it.